[RFC, WIP] Unfungle: Unroll Deps, mypy

crutcher opened the pull request:
This is not meant to be submitted. This is a research PR to work out type/import ordering.

This is a first-pass at trying to flatten and order the import deps, and get mypy to eat them, per issue:
https://github.com/openai/triton/issues/884

most of the contents of `triton` have been unrolled into `triton.unroll`

the tests in `python/tests/unit/language` pass, with the exception of the libdevice ones (I've got the LD_PATH wrong here); and this passes mypy (on `triton.unroll`) except for the block type handling of pointer types in `_save()` and `_load()`; and that's mainly because I don't understand the full indirection pattern there.

Further pull on this would incrementally re-roll back into submodules; upon finding a legal module order and type annotations, we could replay that refactor on the `triton-mlir` branch when it lands.

Jokeren commented:
Do you have any mypy config file?

crutcher commented:
I was using the defaults on this. If this was a real "lets add mypy" PR, we'd need to talk about project structure. Mypy wants to search for its config file in the project root directory; but the project as currently structured places most of the python packaging in a sub-folder; so we could keep that and put the configs in setup.cfg; or migrate the project to pyproject.toml; but we could also move the python packaging to the root of the repo, and adjust the relative paths to things; idk

the mypy defaults are reasonable; but in a maturely typed project, we can ask mypy to be quite a bit more strict (at the cost of refactoring to meet those needs), and get additional safety and refactoring help from that.

the biggest wins though are associated in attaching overrides in the processing of imported libs which don't have correct typing information yet; to suppress the errors that come from them; that's a definite win to do.

crutcher commented:
This is a reasonably complete demonstration of a re-roll, with heavy mypy type additions.

Tests pass, except for some issues with NaNs, which may be a difference between a vt100 and an rtx3090; and libdevice.so.10 import is broken for me and I've not debugged that.

There are two large remaining categories of mypy errors:
 * dtype coercion in the pointer redirect code isn't done, because I don't entirely understand that path and want to ask some questions.
 * tensor.__<op>__ types break the signature; this could be `# type: ignore`d

Here's the mypy log:
```
/home/crutcher/git/third-party/triton/.venv/bin/python -m mypy --config-file python/setup.cfg python/triton 
python/triton/ops/blocksparse/matmul.py:82: error: Unsupported operand type for unary - ("constexpr")  [operator]
python/triton/ops/blocksparse/matmul.py:249: error: No overload variant of "range" matches argument types "tensor", "int", "Any"  [call-overload]
python/triton/ops/blocksparse/matmul.py:249: note: Possible overload variants:
python/triton/ops/blocksparse/matmul.py:249: note:     def range(self, SupportsIndex, /) -> range
python/triton/ops/blocksparse/matmul.py:249: note:     def range(self, SupportsIndex, SupportsIndex, SupportsIndex = ..., /) -> range
python/triton/ops/blocksparse/matmul.py:249: error: Unsupported operand type for unary - ("constexpr")  [operator]
python/triton/core.py:88: error: Cannot determine type of "float16"  [has-type]
python/triton/core.py:606: error: Incompatible default for argument "ieee_rounding" (default has type "bool", argument has type "constexpr")  [assignment]
python/triton/base.py:490: note: "__add__" of "tensor" defined here
python/triton/base.py:495: error: Unexpected keyword argument "_builder" for "__add__" of "tensor"  [call-arg]
python/triton/base.py:507: note: "__mul__" of "tensor" defined here
python/triton/base.py:512: error: Unexpected keyword argument "_builder" for "__mul__" of "tensor"  [call-arg]
python/triton/base.py:627: error: Signature of "__eq__" incompatible with supertype "object"  [override]
python/triton/base.py:627: note:      Superclass:
python/triton/base.py:627: note:          def __eq__(self, object, /) -> bool
python/triton/base.py:627: note:      Subclass:
python/triton/base.py:627: note:          def __eq__(self, Any, Any = ..., /) -> tensor
python/triton/base.py:632: error: Signature of "__ne__" incompatible with supertype "object"  [override]
python/triton/base.py:632: note:      Superclass:
python/triton/base.py:632: note:          def __ne__(self, object, /) -> bool
python/triton/base.py:632: note:      Subclass:
python/triton/base.py:632: note:          def __ne__(self, Any, Any = ..., /) -> tensor
python/triton/base.py:1590: error: "dtype" has no attribute "get_block_shapes"  [attr-defined]
python/triton/base.py:1592: error: "dtype" has no attribute "get_block_shapes"  [attr-defined]
python/triton/base.py:1595: error: "dtype" has no attribute "element_ty"  [attr-defined]
python/triton/base.py:1597: error: "dtype" has no attribute "element_ty"  [attr-defined]
python/triton/base.py:1601: error: "dtype" has no attribute "address_space"  [attr-defined]
python/triton/base.py:1618: error: "dtype" has no attribute "get_block_shapes"  [attr-defined]
python/triton/base.py:1633: error: "dtype" has no attribute "get_block_shapes"  [attr-defined]
python/triton/base.py:1656: error: "dtype" has no attribute "get_block_shapes"  [attr-defined]
python/triton/base.py:1658: error: "dtype" has no attribute "get_block_shapes"  [attr-defined]
python/triton/base.py:1660: error: "dtype" has no attribute "element_ty"  [attr-defined]
python/triton/base.py:1666: error: "dtype" has no attribute "address_space"  [attr-defined]
python/triton/base.py:1687: error: "dtype" has no attribute "element_ty"  [attr-defined]
python/triton/base.py:1703: error: "dtype" has no attribute "element_ty"  [attr-defined]
python/triton/base.py:1710: error: "dtype" has no attribute "get_block_shapes"  [attr-defined]
python/triton/base.py:1712: error: "dtype" has no attribute "get_block_shapes"  [attr-defined]
python/triton/base.py:1713: error: "dtype" has no attribute "element_ty"  [attr-defined]
python/triton/base.py:1718: error: "dtype" has no attribute "get_block_shapes"  [attr-defined]
python/triton/base.py:1719: error: "dtype" has no attribute "get_block_shapes"  [attr-defined]
python/triton/ops/matmul.py:201: error: Unsupported operand type for unary - ("constexpr")  [operator]
python/triton/ops/matmul.py:209: error: Unsupported left operand type for * ("constexpr")  [operator]
python/triton/ops/matmul.py:210: error: Unsupported left operand type for * ("constexpr")  [operator]
Found 30 errors in 4 files (checked 24 source files)
```

crutcher commented:
Also, I ran `black` to maintain sanity while things got smashed about

ptillet commented:
Thanks! Before we look at it, would it be possible to revert the commit that uses `black`? We don't use it (yet) but it definitely makes the diff bigger than it should be.

I also doubt the nans are due to RTX3090 -- it's probably worth investigating further

ptillet commented:
I should also say that the goal of the triton-mlir merge is to have a minimal diff that targets mostly the backend and the IR code generation. Also, changing the namespace for some utilities will likely break some user code. For these reasons, I think we should hold off this refactor until the triton-mlir branch is merged -- probably in a month or so

crutcher commented:
When I was first unrolling, I collapsed everything into one namespace; and ran into conflicts with `cdiv` before i understood the situation; now that everything is restored to `triton` and `triton.language`, the `triton.cdiv` can be returned.

This diff can serve as a guide to a series of smaller incremental diffs replayed on top of that branch once it's merged; we can take the order in this diff, and incrementally constructing it from the top of the import order on a new branch

Jokeren commented:
I'm concerned about the changes in indentation and spacing. Is it caused by `black`?

crutcher commented:
lifted the implementation core into `triton.impl`, pulled all non-core implementations down into `triton.language.stdlib`; cleaned up a bunch of mypy stuff, fixed `version_key()`, added stubs to replicate the old import locations, unified `builtin()/extern()`
